### ISSUE TITLE
Successful termination when constraints violated

### ISSUE THREAD

**Issue by stkiese** [xypron/jcobyla#1]
I have a similar issue as described in https://github.com/scipy/scipy/issues/2891 for the python port of cobyla2: the minimization can terminate without actually satisfying the constraints.

I've uploaded an example to pastebin: http://pastebin.com/p61JmLsb

In my case, reducing rhobeg (solution in https://github.com/scipy/scipy/issues/2891 ) won't help. I've tried different values for rhobeg:

rhobeg=1 => MAXCV=0.091
rhobeg=0.75 => MAXCV=0.0543
rhobeg=0.5 => MAXCV=0.0287
rhobeg=0.25 => MAXCV=0.0686
rhobeg=0.1 => MAXCV=0.0925
rhobeg=0.05 => MAXCV=0.0735
rhobeg=0.001 => MAXCV=0.254

It seems that the

6th (the 6th input var has to be greater or equal to zero), 
(9th or 10th) (the sum of all input vars has to be approximately 1)
and 11st (to complex to explain in one sentence ;-) )

constraints are violated. But there is a solution that satisfies all constraints:

{0.705, 0.0075, 0.075, 0, 0.0375, 0, 0.05, 0.125}

But even if I start with this vector, I have to decrease rhobeg to 0.0005 to get a solution with a violation level lower than 0.001.

Is there a bug in my code, in jcobyla, a bug in the java port of cobyla or just "normal behavior"? 

For the sake of completeness: I've also tried another java cobyla implementation, https://github.com/cureos/jcobyla . The cureos implentation exhibits the same behavior. The author mentioned that it could be that cobyla has problems with equality constraints in general: https://github.com/cureos/jcobyla/issues/1

Thanks!
Stefan

**Comment by xypron** [comment]
Hello Stefan,

On 16.01.2015 07:35, stkiese wrote:

> I have a similar issue as described in scipy/scipy#2891
> https://github.com/scipy/scipy/issues/2891 for the python port of
> cobyla2: the minimization can terminate without actually satisfying the
> constraints.
> 
> I've uploaded an example to pastebin: http://pastebin.com/p61JmLsb
> 
> In my case, reducing rhobeg (solution in scipy/scipy#2891
> https://github.com/scipy/scipy/issues/2891 ) won't help. I've tried
> different values for rhobeg:
> 
> rhobeg=1 => MAXCV=0.091
> rhobeg=0.75 => MAXCV=0.0543
> rhobeg=0.5 => MAXCV=0.0287
> rhobeg=0.25 => MAXCV=0.0686
> rhobeg=0.1 => MAXCV=0.0925
> rhobeg=0.05 => MAXCV=0.0735
> rhobeg=0.001 => MAXCV=0.254
> 
> It seems that the
> 
> 6th (the 6th input var has to be greater or equal to zero),
> (9th or 10th) (the sum of all input vars has to be approximately 1)
> and 11st (to complex to explain in one sentence ;-) )
> 
> constraints are violated. But there is a solution that satisfies all
> constraints:
> 
> {0.705, 0.0075, 0.075, 0, 0.0375, 0, 0.05, 0.125}
> 
> But even if I start with this vector, I have to decrease rhobeg to
> 0.0005 to get a solution with a violation level lower than 0.001.

Rhobeg defines the initial size of the simplex with which the search begins.

Please, read Powell's article for details of the algorithm

```
M. J. D. Powell, "A direct search optimization method that
models the objective and constraint functions by linear
interpolation," in Advances in Optimization and Numerical
Analysis, eds. S. Gomez and J.-P. Hennart (Kluwer Academic:
Dordrecht, 1994), p. 51-67.
```

and

http://www.damtp.cam.ac.uk/user/na/NA_papers/NA2007_03.pdf

> Is there a bug in my code, in jcobyla, a bug in the java port of cobyla
> or just "normal behavior"?
> 
> For the sake of completeness: I've also tried another java cobyla
> implementation, https://github.com/cureos/jcobyla . The cureos

I simply copied cureos code and made it compatible with Maven.
No modification of the algorithm was implemented by me.

A bunch of different optimization algorithms are implemented in NLopt.
Amongst them cobyla.

http://ab-initio.mit.edu/wiki/index.php/NLopt_release_notes
talks about several improvements to the cobyla algorithm.

The code is available at
https://github.com/stevengj/nlopt/tree/master/cobyla

So maybe it is worthwhile to transfer these changes to jcobyla.

I would be happy to add your patch and upload to the Maven repository.

> implentation exhibits the same behavior. The author mentioned that it
> could be that cobyla has problems with equality constraints in general:
> cureos/jcobyla#1 https://github.com/cureos/jcobyla/issues/1

You can remove the equality
sum(x) = 1
by the following substitution removing one variable:

x8 = 1 - x1 - x2 - x3 - x4 - x5 - x6 - x7

> Thanks!
> Stefan
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/xypron/jcobyla/issues/1.

Best regards

Heinrich Schuchardt

mailto:xypron.glpk@gmx.de

**Comment by xypron** [comment]
The code comments say:
"constraint functions should become nonnegative eventually, at least to the precision of RHOEND."

Is the violoation exceeding RHOEND in your case? If yes, this would be an error.

MAXCV is not an absolute measure of the constraint violation, but indicates a multiple of the stepsize. So as long as the value does not exceed 1.0 everything should be ok.

**Comment by stkiese** [comment]
Hello Heinrich,

thank you for your fast reply! I'm going to read through these articles, but this could take a while.

I also tested my code with the nlopt C# wrapper NLoptNet ( https://github.com/BrannonKing/NLoptNet ). Also the NLoptNet Cobyla implementation isn't able to solve my problem. In difference to your implementation, NLoptNet returns "Infinity" as result. This could be a hint that the optimization has failed. JCobyla returns the function value as result.

I would like to port the Cobyla nlopt improvements to jcobyla, but I don't think I have the mathematical understanding to do so :-( .

Thanks for the hint how to remove the equality constraint. Unfortunately this won't really improve the result.

http://pastebin.com/CmPEza1E

Yes, I've found this comment as well and regarding to this comment my issue could be a bug in cobyla (or the jcobyla port, or I use jcobyla in a wrong way). There is another comment that says

"Therefore RHOBEG and  RHOEND should be set to reasonable initial changes to and the required  accuracy in the variables respectively, but this accuracy should be viewed as a subject for experimentation because it is not guaranteed." (cobyla2.f, line 16-19)

I haven't figured out whether this comment means that cobyla isn't able to find the minimum and respects the constraints with the required accuracy in any case or not.

Yes, the violation is exceeding rhoend clearly. The rhoend value is set to 1.E-6, the violation of at least one constraint is between 0.02 and 0.09 depending on the chosen rhobeg. Only if I start with a vector that doesn't violate the constraints (the solution that jcobyla is supposed to find) and if I set rhobeg below 0.001, the violation is acceptable.

I have checked the violation of the constraint after jcobyla has finished its calculation. MAXCV is exactly Min(0, con1, ..., con11) \* -1. MAXCV doesn't exceed 1.0 but I don't think that everything is ok, because the constraints are violated clearly above rhoend.

Thanks
Stefan
